pkg/kubelet/deviceplugin/manager.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deviceplugin
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha1"
+)
+
+// NewManagerImpl creates a new manager on the socket `socketPath` and can
+// rebuild state from devices and available []Device.
+// f is the callback that is called when a device becomes unhealthy
+// socketPath is present for testing purposes in production this is pluginapi.KubeletSocket
+func NewManagerImpl(socketPath string, f MonitorCallback) (*ManagerImpl, error) {
+	glog.V(2).Infof("Creating Device Plugin manager at %s", socketPath)
+
+	if socketPath == "" || !filepath.IsAbs(socketPath) {
+		return nil, fmt.Errorf(ErrBadSocket+" %v", socketPath)
+	}
+
+	dir, file := filepath.Split(socketPath)
+	return &ManagerImpl{
+		Endpoints: make(map[string]*endpoint),
+
+		socketname: file,
+		socketdir:  dir,
+		callback:   f,
+	}, nil
+}
+
+// Start starts the Device Plugin Manager
+func (m *ManagerImpl) Start() error {
+	glog.V(2).Infof("Starting Device Plugin manager")
+
+	socketPath := filepath.Join(m.socketdir, m.socketname)
+	os.MkdirAll(m.socketdir, 0755)
+
+	if err := os.Remove(socketPath); err != nil && !os.IsNotExist(err) {
+		glog.Errorf(ErrRemoveSocket+" %+v", err)
+		return err
+	}
+
+	s, err := net.Listen("unix", socketPath)
+	if err != nil {
+		glog.Errorf(ErrListenSocket+" %+v", err)
+		return err
+	}
+
+	m.server = grpc.NewServer([]grpc.ServerOption{}...)
+
+	pluginapi.RegisterRegistrationServer(m.server, m)
+	go m.server.Serve(s)
+
+	return nil
+}
+
+// Devices is the map of devices that are known by the Device
+// Plugin manager with the Kind of the devices as key
+func (m *ManagerImpl) Devices() map[string][]*pluginapi.Device {
+	glog.V(2).Infof("Devices called")
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	devs := make(map[string][]*pluginapi.Device)
+	for k, e := range m.Endpoints {
+		glog.V(2).Infof("Endpoint: %+v: %+v", k, e)
+		e.mutex.Lock()
```

Comment:
The first lock ensures that the `endpoints` map won't be changing underneath us.

The second one ensures that the `devices` map in an `Endpoint` won't be changing underneath us either.

The model of having an Endpoint in charge of it's map of Device was an idea from @jiayingz in order to prevent a device plugin from blocking other device plugins with too many Watch requests.
It also means that we need two mutex